albogdan/skule-vote
# Submit Ballot API Endpoint

## Overview

- Resolves #6 
- Added functionality to handle ballot submissions
   - BallotSubmitView checks the following eligibility criteria:
     - Voter is logged in
     - The election referenced exists
     - Voter is eligible to vote in that particular election
     - Voter has not already voted in that particular election
   - BallotSerializer is responsible for converting the json payload from frontend into Ballot models. It ensures all candidates referenced are valid and then saves the models inside an atomic transaction.
     - Payload structure is based on what is currently in frontend 


## Unit Tests Created

- Added unit tests to test various cases of ineligible voters trying to submit ballots:
  - People who haven't logged in
  - People who have tried to modify their cookie
  - People who are trying to vote in another year/discipline's election
  - People who are full time but are voting for part time chair
  - People who are trying to vote twice in an election they are eligible in
- Also added unit tests to test successful submission of ballots in different eligibility paths, including a specific test for spoiled ballots


## Steps to QA

- Create an election session with a few elections and add candidates.
- Use the cookie bypass view to log in
- Use the swagger view to send post ballot requests
- Make sure the ballots in the DB are what they should be
- Try to stress test eligibility criteria
